keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,294 @@
+events to the watch automatically. By default etcd sends progress events every
+10 minutes, which is not frequent enough to be useful for our needs, so we will
+modify etcd to send them must more frequently, e.g. every 20ms.
+
+When an consistent read request is received and the watch cache is enabled:
+
+- Get the current revision from etcd for the resource type being served. The returned revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly (20ms?) for the watch to catch up to the current revision.
+- If the block times out, skip the cache and serve the request directly from storage (etcd).
+
+To get the revsion we have two options: 
+
+- Use an etcd range request with `WithCount` enabled so etcd return only a count and revision 
+- Use an etcd range request against a known empty range with limit=1 as an additional guard (since etcd does not allow for limit=0)
+
+We will benchmark and select the most efficient of these options.
+
+#### Alternative 2: Use WatchProgressRequest to request watch updates when needed
+
+etcd 3.4+ provides a `WatchProgressRequest` request that can be made on a watch channel. When requested,
+etcd will send a progress event on that watch as soon as possible.
+
+When an consistent read request is received and the watch cache is enabled:
+- Get the current revision from etcd using a range read with limit=0, just like in alternative 1.
+- If the cache already has the current revision, serve the request from cache
+- If the current revision is not in the cache:
+  - Send a `WatchProgressRequest` to etcd on the watch channel that the watch cache is consuming.
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
+- If the block times out, skip the cache and serve the request directly from storage.
+
+This alternative requires using `WatchProgressRequest` which is only available in etcd 3.4+, and so would
+require we make the kube-apiserver aware of etcd's minor version, which is described in more detail later.
+
+#### Comparing the alternatives
+
+We will experiment with modifying etcd to automatically send progress events in
+short intervals of 20ms or so to see if this keeps the latest observed revision
+of watch caches sufficiently up-to-date that they rarely need to fall back to
+serving resourceVersion="" requests directly from etcd.  We will check for cache
+hit ratios and check scalability. If these are promising we will then explore
+what the optimal interval to have etcd send progress events is. If this works
+well enough we will go with alternative 1 and will not need to implement
+alternative 2.
+
+We will also explore how long the kube-apiserver should wait for a watch cache
+to catch up to the needed revision. I.e. for a progress event interval of 20ms,
```

Comment:
If people want minimal latency, they can provide an RV, no?

I think it's pretty reasonable to make them wait an extra 500ms for consistent reads; so I think 250ms (to allow for a near miss) might be a more reasonable number than 20ms.

Reply:
Sounds good. @mm4tt ran some experiments (https://github.com/kubernetes/kubernetes/pull/86769) and a progress interval of 50ms did result in noticeable performance impact, so next we can do a run with 250ms and see how that performs.

Reply:
We could even go to 333ms, but not all the way to 500ms, I think--we wouldn't want 2 of those to miss. I think we should try and keep the overall latency below a second.

Reply:
250ms did fine in the last scalability test run. We'll go with that in the doc for now.